File: python/mahjong_utils/shanten.py
from enum import Enum
from typing import Optional, Sequence, Set, Dict

from pydantic import BaseModel
from stringcase import snakecase

from mahjong_utils.lib import libmahjongutils
from mahjong_utils.models.furo import Furo
from mahjong_utils.models.hand import Hand
from mahjong_utils.models.shanten import Shanten, ShantenWithoutGot
from mahjong_utils.models.tile import Tile


class ShantenResultType(str, Enum):
    regular = "Regular"
    chitoi = "Chitoi"
    kokushi = "Kokushi"
    union = "Union"


class ShantenResult(BaseModel):
    type: ShantenResultType
    hand: Hand
    shanten_info: Shanten
    regular: Optional["ShantenResult"]
    chitoi: Optional["ShantenResult"]
    kokushi: Optional["ShantenResult"]

    def encode(self) -> dict:
        return dict(
            type=self.type.name,
            hand=self.hand.__encode__(),
            shantenInfo=self.shanten_info.__encode__(),
            regular=regular.encode() if (regular := self.regular) is not None else None,
            chitoi=chitoi.encode() if (chitoi := self.chitoi) is not None else None,
            kokushi=kokushi.encode() if (kokushi := self.kokushi) is not None else None,
        )

    @classmethod
    def decode(cls, data: dict) -> "ShantenResult":
        return ShantenResult(
            type=ShantenResultType[snakecase(data["type"])],
            hand=Hand.__decode__(data["hand"]),
            shanten_info=Shanten.__decode__(data["shantenInfo"]),
            regular=ShantenResult.decode(regular) if (regular := data["regular"]) is not None else None,
            chitoi=ShantenResult.decode(chitoi) if (chitoi := data["chitoi"]) is not None else None,
            kokushi=ShantenResult.decode(kokushi) if (kokushi := data["kokushi"]) is not None else None,
        )

    @property
    def shanten(self) -> Optional[int]:
        return getattr(self.shanten_info, "shanten", None)

    @property
    def advance(self) -> Optional[Set[Tile]]:
        return getattr(self.shanten_info, "advance", None)

    @property
    def advance_num(self) -> Optional[int]:
        return getattr(self.shanten_info, "advance_num", None)

    @property
    def well_shape_advance(self) -> Optional[Set[Tile]]:
        return getattr(self.shanten_info, "well_shape_advance", None)

    @property
    def well_shape_advance_num(self) -> Optional[int]:
        return getattr(self.shanten_info, "well_shape_advance_num", None)

    @property
    def discard_to_advance(self) -> Optional[Dict[Tile, ShantenWithoutGot]]:
        return getattr(self.shanten_info, "discard_to_advance", None)

    @property
    def with_got(self) -> bool:
        return self.discard_to_advance is not None


def regular_shanten(
        tiles: Sequence[Tile],
        furo: Optional[Sequence[Furo]] = None,
        calc_advance_num: bool = True
) -> ShantenResult:
    result = libmahjongutils.call("regularShanten", {
        "tiles": [str(t) for t in tiles],
        "furo": [Furo.__encode__(fr) for fr in furo] if furo is not None else [],
        "calcAdvanceNum": calc_advance_num
    })

    return ShantenResult.decode(result)


def chitoi_shanten(
        tiles: Sequence[Tile],
        calc_advance_num: bool = True
) -> ShantenResult:
    result = libmahjongutils.call("chitoiShanten", {
        "tiles": [str(t) for t in tiles],
        "calcAdvanceNum": calc_advance_num
    })

    return ShantenResult.decode(result)


def kokushi_shanten(
        tiles: Sequence[Tile],
        calc_advance_num: bool = True
) -> ShantenResult:
    result = libmahjongutils.call("kokushiShanten", {
        "tiles": [str(t) for t in tiles],
        "calcAdvanceNum": calc_advance_num
    })

    return ShantenResult.decode(result)


def shanten(
        tiles: Sequence[Tile],
        furo: Optional[Sequence[Furo]] = None,
        calc_advance_num: bool = True
) -> ShantenResult:
    result = libmahjongutils.call("shanten", {
        "tiles": [str(t) for t in tiles],
        "furo": [Furo.__encode__(fr) for fr in furo] if furo is not None else [],
        "calcAdvanceNum": calc_advance_num
    })

    return ShantenResult.decode(result)


__all__ = ("regular_shanten",
           "chitoi_shanten",
           "kokushi_shanten",
           "shanten",
           "ShantenResult",)

File: mahjong-utils/src/commonMain/kotlin/mahjongutils/shanten/Shanten.kt
package mahjongutils.shanten

import mahjongutils.CalcContext
import mahjongutils.models.Furo
import mahjongutils.models.Tile
import mahjongutils.models.countAsCodeArray
import mahjongutils.models.hand.CommonHandPattern
import mahjongutils.models.hand.Hand
import kotlin.jvm.JvmOverloads
import kotlin.math.min

/**
 * 向听分析参数（内部计算使用）
 * @param tiles 门前的牌
 * @param furo 副露（对向听分析本身无用，但若需要将结果用于和了分析则需要传入）
 * @param calcAdvanceNum 是否计算进张数
 * @param calcGoodShapeAdvance 是否计算好型进张
 * @param bestShantenOnly 仅计算最优向听数的打法（不计算退向打法）
 * @param allowAnkan 是否允许暗杠
 * @param calcImprovement 是否计算改良
 * @return 向听分析结果
 */
internal data class InternalShantenArgs(
    val tiles: List<Tile>,
    val furo: List<Furo> = emptyList(),
    val calcAdvanceNum: Boolean = true,
    val calcGoodShapeAdvance: Boolean = true,
    val bestShantenOnly: Boolean = false,
    val allowAnkan: Boolean = true,
    val calcImprovement: Boolean = true,
)

/**
 * 向听分析
 * @param tiles 门前的牌
 * @param furo 副露（对向听分析本身无用，但若需要将结果用于和了分析则需要传入）
 * @param bestShantenOnly 仅计算最优向听数的打法（不计算退向打法）
 * @return 向听分析结果
 */
@JvmOverloads
fun shanten(
    tiles: List<Tile>,
    furo: List<Furo> = emptyList(),
    bestShantenOnly: Boolean = false,
): UnionShantenResult {
    val internalShantenArgs = InternalShantenArgs(
        tiles = tiles,
        furo = furo,
        bestShantenOnly = bestShantenOnly
    )

    val context = CalcContext()
    return context.shanten(internalShantenArgs)
}

/**
 * 向听分析
 * @param args 向听分析参数
 * @return 向听分析结果
 */
fun shanten(
    args: ShantenArgs
): UnionShantenResult {
    val internalShantenArgs = InternalShantenArgs(
        tiles = args.tiles,
        furo = args.furo,
        bestShantenOnly = args.bestShantenOnly
    )

    val context = CalcContext()
    return context.shanten(internalShantenArgs)
}


internal fun CalcContext.shanten(
    args: InternalShantenArgs
): UnionShantenResult = memo(Pair("shanten", args)) {
    with(args) {
        val tiles = ensureLegalTiles(tiles)

        val withGot = tiles.size % 3 == 2
        val k = tiles.size / 3

        if (k != 4) {
            val regular = regularShanten(
                copy(tiles = tiles)
            )
            return UnionShantenResult(
                hand = regular.hand, shantenInfo = regular.shantenInfo,
                regular = regular
            )
        }

        val regular = regularShanten(
            copy(tiles = tiles, calcAdvanceNum = false)
        )
        val chitoi = chitoiShanten(
            InternalShantenArgs(tiles = tiles, calcAdvanceNum = false, bestShantenOnly = bestShantenOnly)
        )
        val kokushi = kokushiShanten(
            InternalShantenArgs(tiles = tiles, calcAdvanceNum = false, bestShantenOnly = bestShantenOnly)
        )

        val shantenNum = min(
            min(
                regular.shantenInfo.shantenNum,
                chitoi.shantenInfo.shantenNum
            ),
            kokushi.shantenInfo.shantenNum
        )
        val patterns = ArrayList<CommonHandPattern>()

        var shantenInfo = if (!withGot) {
            val advance = HashSet<Tile>()
            val goodShapeAdvance = HashSet<Tile>()

            mergeIntoWithoutGot(shantenNum, advance, goodShapeAdvance, patterns, regular)
            mergeIntoWithoutGot(shantenNum, advance, goodShapeAdvance, patterns, chitoi)
            mergeIntoWithoutGot(shantenNum, advance, goodShapeAdvance, patterns, kokushi)

            ShantenWithoutGot(
                shantenNum, advance,
                goodShapeAdvance = if (shantenNum == 1) goodShapeAdvance else null,
                improvement = (regular.shantenInfo.asWithoutGot).improvement
                    ?: (if (shantenNum == 0) emptyMap() else null),
                goodShapeImprovement = (regular.shantenInfo.asWithoutGot).goodShapeImprovement
                    ?: (if (shantenNum == 0) emptyMap() else null)
            )
        } else {
            val discardToAdvance = HashMap<Tile, ShantenWithoutGot>()

            mergeIntoWithGot(shantenNum, discardToAdvance, patterns, regular, bestShantenOnly)
            mergeIntoWithGot(shantenNum, discardToAdvance, patterns, chitoi, bestShantenOnly)
            mergeIntoWithGot(shantenNum, discardToAdvance, patterns, kokushi, bestShantenOnly)

            ShantenWithGot(
                shantenNum, discardToAdvance,
                ankanToAdvance = (regular.shantenInfo.asWithGot).ankanToAdvance
            )
        }

        if (calcAdvanceNum) {
            val tilesCount = (tiles + furo.flatMap { it.asMentsu().tiles }).countAsCodeArray()
            shantenInfo = shantenInfo.fillNum(tilesCount)
        }

        val hand = Hand(tiles = tiles, furo = furo, patterns = patterns)
        return UnionShantenResult(
            hand = hand, shantenInfo = shantenInfo,
            regular = regular, chitoi = chitoi, kokushi = kokushi
        )
    }
}


private fun <S : CommonShanten> mergeIntoWithoutGot(
    targetShantenNum: Int,
    advance: MutableSet<Tile>,
    goodShapeAdvance: MutableSet<Tile>,
    patterns: MutableCollection<CommonHandPattern>,
    result: ShantenResult<S, *>
) {
    if (result.shantenInfo.shantenNum == targetShantenNum) {
        val shantenInfo = result.shantenInfo.asWithoutGot
        advance += shantenInfo.advance
        patterns += result.hand.patterns
        if (shantenInfo.goodShapeAdvance != null) {
            goodShapeAdvance += shantenInfo.goodShapeAdvance
        }
    }
}

private fun <S : CommonShanten> mergeIntoWithGot(
    targetShantenNum: Int,
    discardToAdvance: MutableMap<Tile, ShantenWithoutGot>,
    patterns: MutableCollection<CommonHandPattern>,
    result: ShantenResult<S, *>,
    bestShantenOnly: Boolean
) {
    val shantenInfo = result.shantenInfo.asWithGot
    for ((discard, shantenAfterDiscard) in shantenInfo.discardToAdvance) {
        if (!bestShantenOnly || shantenAfterDiscard.shantenNum == targetShantenNum) {
            val oldShantenAfterDiscard = discardToAdvance[discard]
            if (oldShantenAfterDiscard == null || oldShantenAfterDiscard.shantenNum > shantenAfterDiscard.shantenNum) {
                discardToAdvance[discard] = shantenAfterDiscard
            } else if (oldShantenAfterDiscard.shantenNum == shantenAfterDiscard.shantenNum) {
                discardToAdvance[discard] = oldShantenAfterDiscard.copy(
                    advance = oldShantenAfterDiscard.advance + shantenAfterDiscard.advance,
                    goodShapeAdvance = shantenAfterDiscard.goodShapeAdvance?.let {
                        oldShantenAfterDiscard.goodShapeAdvance?.plus(it)
                    }
                )
            }
        }
    }

    if (targetShantenNum == shantenInfo.shantenNum) {
        patterns += result.hand.patterns
    }
}